### ISSUE TITLE
Nodes with library styles gets their selected style overwritten with hex value

### ISSUE THREAD

**Issue by jakobe** [six7/figma-tokens#758]
**Describe the bug**
When a node in the document is attached to both a Figma Tokens plugin token (e.g. `color.core.gray.50`) and gets it's color from a library style (e.g. `color/core/gray/50`) the Figma color style is replaced with the hex value from the token.

This happens both when explicitly assigning a token that corresponds to a Figma color style from a library but also on overall updates from the plugin (e.g. when toggling token sets in the plugin).

The bug also surfaces when copying an element (e.g. a component) from the library master file where the color styles are defined and that element has a color token assigned to it. When pasting that element (e.g. as an instance) into another file that uses the color styles from the library the color styles are overwritten with hex values on updates from the plugin (e.g. when toggling token sets in the plugin).

_*Just for the record: Text and effect styles are also affected by this issue, I'm just using colors as an example here_

**To Reproduce**
1. Create a new file in Figma
2. Add a new token set (e.g. `Global`) in the Figma Tokens plugin
3. Add a color token (e.g. `color.core.gray.50`) in the plugin and create Figma color styles from the tokens (`Styles > Create styles`)
4. Publish the file as a library
5. Create a new file and enable the published library from step #&#x2060;4
6.  Create a rectangle and assign the color token created in step #&#x2060;3 (e.g. `color.core.gray.50`) through the plugin corresponding to the Figma library color style (`color/core/gray/50`)
7. See that the hex value is applied (e.g. `F6F6F6`) and not the Figma library color style (`color/core/gray/50`)
8. Assign the Figma library color style (`color/core/gray/50`) through Figma's `Design` panel.
9. Assign the color token created in step #&#x2060;3 (e.g. `color.core.gray.50`) through the plugin
10. See that the hex value is applied (e.g. `F6F6F6`) and not the Figma library color style (`color/core/gray/50`)

**Expected behavior**
I would expect the plugin to NOT overwrite my applied Figma color styles.  
Either because it can map the color token to the - non-local - color styles from the library OR  
when it cannot because it sees a color style is applied and doesn't overwrite it.

I guess the main reason for this bug is because [Figma only exposes the local paint styles](https://www.figma.com/plugin-docs/api/figma/#getlocalpaintstyles) and [does not (yet at least) expose a way to get library styles ](https://forum.figma.com/t/getlocalpaintstyles-but-for-libraries/1280)

@six7 As far as I can see the offending lines are in [src/plugin/setValuesOnNode.ts](https://github.com/six7/figma-tokens/blob/main/src/plugin/setValuesOnNode.ts#L82-L88).
If you agree on the expected behavior, maybe a simply check on existing `node.fillStyleId` before assigning would do the trick?
We could even extend it further with a check for whether the style exists (`figma.getStyleById(node.fillStyleId)`) and add that style to the `figmaStyleMaps` used to check for matching styles?

I'll be happy to do a fix and PR if you see fit 🤞🏻 😊

**Comment by six7** [comment]
👋 its a little more complicated than that, but we're working on this already!

Basically you cannot expect just because a token has the same name as an applied style that applying that style is the right thing. If using option tokens and the color value matches between style and token (like the one in your example) then yes, keeping that style is the way to go.

However when using Theming and your token values change from light to dark for a bg.default you need to swap that style for the corresponding style requiring us to keep track of what token/theme combination matches what style.

In addition, ideally it not just keeps a style, but is able to apply a style rather than a hex value when updating - requiring each token to either have a theme level style id representation or on a per token basis - we're currently working on first the theming part and later on the individual part 🎉

Also, this issue is already covered by https://github.com/six7/figma-tokens/issues/145

**Comment by jakobe** [comment]
👋🏻  again,

Sorry I didn't see #145 (I _did_ look through all issues first to see if I could find a similar one, but it must've slipped my eyes 🙈 )

I follow what you're saying above ☝🏻 , nevertheless this is exactly what the plugin is doing at the moment - i.e. assuming that if the name of the token with `.` as a delimiter matches a local Figma style with `/`as the delimiter it should assign that Figma style (and not the value of the plugin token). It doesn't check the current value if the node at all.
It also DOES apply the style and not the hex value, if the style is found using the mentioned naming scheme (so not sure what you mean by `ideally it not just keeps a style, but is able to apply a style rather than a hex value when updating`? 🤔 )

Not sure I follow the Theming part though? If you mean "Theming" as in "Token sets", then as far as I know the tokens doesn't change their name or nesting, only the value changes (which is the point of Theming anyways 😀 )
And in the linked code lines and as explained above it doesn't check whether the value matches (which I then WOULD be a dealbreaker in regards to Theming).

TL;DR: Great you're already working on the Theming part and are looking to solve the individual part as well 🎉 
Until that settles though, it would be a great help, if the plugin doesn't remove / overwrite applied Figma color styles with token values when it doesn't find a matching Figma style *when the node already has an applied Figma style* (emphasis is on this last part, I guess I didn't make that clear enough).

**Comment by six7** [comment]
Yeah I think that makes sense! However let's only do that as long as the style value and the token value match (name alone isn't enough). Should be easy to compare, would just need to make sure it doesn't add a performance overhead.

Also, if possible base the PR on the `next` branch as we reserve the main branch for whatever's in prod right now (and as we have no way of doing automatic releases with Figma it's a manual process).

**Comment by six7** [comment]
And let me give you an example on the token sets:

[token set light]: bg.default = gray.100
[token set dark]: bg.default = gray.900

Style applied on layer: bg.default - if we only went by name, the layer wouldnt update when switching from light to dark. And thats just 1 abstraction. Themes can be very complex and depend on multiple prepending themes, so it's that unique combination that makes a 'Style' (when a user creates library sets they usually activate the required sets)

**Comment by jakobe** [comment]
Hi @six7,

I did a quick draft PR to make sure where on the same page 😊 
I.e. I haven't added any tests (yet), but here it is - based off the `next` branch:
https://github.com/six7/figma-tokens/pull/765

The PR only fixes this ☝🏻  issue (#758), i.e. it ensures that when we cannot match a Figma style (as in the current implementation of that check) we only apply the raw token value to the node *if it doesn't have a linked style already* (i.e. `fillStyleId`, `strokeStyleId`, `textStyleId` or `effectStyleId` is empty).

I'm emphasizing this, as I still can't follow along what you're saying above, sorry 🙈 😄 
Maybe I'm missing something? Or maybe you're talking about how to fix #145 🤷🏻 

Hope this makes sense 🤞🏻

**Comment by six7** [comment]
Let me try to explain again :D

With Figma Tokens, users can have token sets, which can be a way to create themes. Usually, token names for a background.default color token is named the same across a Light and a Dark theme. Users in Figma typically have the same style names, but then just replace the whole library with another one. So the theme name is almost never part of the style name.

Now with your implementation, if there's already a `background.default` style attached with value: black, and I in my plugin change to the Light theme (which should give background a white value), it wouldn't do it because the name of the style is 1:1 the same `background.default`

**Comment by jakobe** [comment]
Hi @six7 ,

Thx for your patience 😊🙏🏻

I pretty sure I get what you’re saying, as we are using Themes/token sets as well, and I see your point.

What I don’t understand though, is that as far as I can see, that path (switching to a new theme/token set with same token names but different values) is already covered in the first `if` that checks for a matching style and applies that, right? My second `if` only makes sure that if it **doesn’t** find a match it doesn’t remove/overwrite the linked style with a raw (token) value.

I’m assuming that shouldn’t be the case either in your example, right? In other words replacing the linked style with a raw value? In your example the first `if` would still assign the linked style and not the raw token value, right?

**Comment by six7** [comment]
In files where no styles are defined, instead getting styles from a library, there currently will never be a matching style. But users still want to switch themes. Which is why we need a check for same values, otherwise switching themes would have no effect.

**Comment by jakobe** [comment]
So if I understand you correctly you **do** want the plugin to remove/overwrite the linked Figma style? I.e. if a user links a node to a style from a library and at the same time attaches a plugin token to that node, then the plugin "wins"?

Normally in Figma you have to detach a style for a node before you can assign a raw hex value - I would expect the plugin to respect that? I.e. if I want to allow the plugin to switch color when switching between tokens sets/themes that are not linked (locally) to my Figma styles then I would need to detach my linked style and use raw values instead.
This hopefully gets better when #145 is fixed, but until then I wouldn't expect the plugin to remove/overwrite the linked Figma style for a node...

**Comment by six7** [comment]
Yes, I do want that as long as the value to be used is different to the Style that is on a layer. Until we have Theming support built in, which would apply the correct style. If we implemented your fix, currently theming would be impossible.

Imagine you created a library file that has styles attached on their layers, as those are local. Now you switch themes, and you notice no change because a style was already attached.

Better outcome: The plugin is instead able to apply the correct style depending on the theme (which is in most cases a different one than the one applied). This requires a style to be set on each token on each theme for complex scenarios, and a style to be defined on a token in simpler scenarios so that the plugin knows what style to apply. Simply not acting and leaving the style that was applied on the layer doesn't allow theming.

I think your solution would work fine as long as we did
- is a style applied on a layer?
- is the style value the same as the value coming in from the token?
  - if yes, don't override with a hex value
  - if no, override with a hex value

**Comment by jakobe** [comment]
Hi @six7 👋🏻, and thanks again for taking your time on this issue 🙏🏻 

Just want to make sure we are talking about the same thing:

>  Imagine you created a library file that has styles attached on their layers, as those are local. Now you switch themes, and you notice no change because a style was already attached.

I assume you mean selecting/deselecting the tabs/token sets in the plugin, right? (Or are you talking about the new theming feature in the pro version of the plugin?)
If the first is the case, then I cannot recreate the issue you're mentioning about changes not being applied?

As far as I can see that still works even with my fix, as the plugin updates the Figma styles and not the raw color of the nodes attached to the token as such (well, it also runs the `setValuesOnNode` function on each node with tokens attached, but that basically just sets `node.fillStyleId` to the matching style, which is already is set to the same `fillStyleId` anyways). 

I've attached a video of me assigning a token (and matching Figma style) to a rectangle node, then selecting a dark theme with a different value for that same token. As you can see the name of the Figma style stays the same, but the value of that style gets updated to the dark theme value.

**Please note** I'm only talking about the scenario where local styles are defined that matches the tokens in the plugin.
For a document with non-local styles it's a different scenario (which this issue is all about 😄 ) 

https://user-images.githubusercontent.com/9556693/170996429-9abd467a-bc1b-42b1-a3f8-a7e75cc41d0d.mov

**Comment by six7** [comment]
I'm mainly talking about the scenario where you created your components in a file that had local styles, and it had them applied (as it was able to find a style with that name).

Then, in another file, you activate a different set. If all we checked was the presence of a style, nothing would change and the original style would persist.

While attaching a hex value is unexpected, so is not changing the style.

As I described, ideally we'd apply a style that matches the token (which we can only if we keep track of style id's). If not, then we should check if the style value === token value

**Comment by six7** [comment]
> Thx @six7, I misunderstood you then regarding the scenario, but get what you're saying now.
> 
> I think what also adds to the confusion is that what I see as a bug you actually see as a feature 😀
> 
> But I get why you want to keep the ability to toggle themes even when a node has a non-local style linked to it (or at least why to keep that feature since users are now used to it working that way).
> 
> I went back to the drawing board and revisited my very first idea of getting and storing the non-local style before overwriting the nodes value with the raw hex value from the token. That way we can restore the linked non-local style whenever the token value matches the linked style value (e.g. when toggling back to default/light theme).
> 
> Since a picture says more than a thousand words and a video multiplies that with FPS 😅 I've attached another video of me toggling a theme/token set with tokens that match both local and non-local styles, showcasing how:
> 
> * local styles gets updated with the token value (current implementation)
> * linked non-local styles being overwritten with raw token value (current implementation)
> * linked non-local styles being re-linked/restored when raw token value matches (new implementation/my fix)
> 
> The WIP changes are in this commit: [jakobe@cf87650](https://github.com/jakobe/figma-tokens/commit/cf876509f278bac8306c9205fec9618239da95a9)
> 
> **Please note** I've only made the changes for `fillStyleId` for now - I will extend this to `strokeStyleId`, `textStyleId` and `effectStyleId` if you agree on this path forward 🤞🏻
> 
>  Figma.Tokens.preserve.non-local.styles.mov

🎉  Works great! There's some thing that we'd need to cover still though - we'd want to do the same thing for all styles, not just colors (effects and type). Also, you're currently comparing only color, however you'd need to take opacity into account as well.

Can you keep the PR on hold for now? We're currently making lots of changes to that logic due to the introduction of the non-native styles for themes - where your PR would fit in nicely for those users who don't want to use the Themes feature. I'll let you know once it's in a state where work can be commenced 👍 

Yeah, its not really a bug, it is a feature as long as we don't have a better way of dealing with Themes. We really can't assume just because a Style has been applied that this is the style the user wants applied due to the way how we're applying tokens as styles on layers just because a local style is found that matches the token name.

**Comment by six7** [comment]
> Hey @six7 , great to see the `next` branch got merged today 🚀 🎉
> 
> I've now cleaned up the code and added checks for non-local styles matching the attached token's (current) value for:
> 
> * fills (`fillStyleId`)
> * borders (`strokeStyleId`)
> * typography (`textStyleId`)
> * effects, i.e. shadows (`effectStyleId`)
> 
> To sum up:
> 
> * When applying a token that do match a local style, the style linked is kept intact and the underlying local Figma style is updated to reflect the current token value(s)
> * When applying a token that doesn't match a local style, and the node is linked to a non-local Figma style, the value(s) of that token will override the linked Figma style when the value(s) doesn't match, e.g. applying a Dark Theme. The id of the non-local style will be backed up (stored on the node as plugin data).
> * When applying an updated token value (i.e. when de-selecting a Dark Theme, going back to Light/Default Theme) where the token value(s) matches the previously backed up non-local style (getting the non-local style using the id from the plugin data) the non-local Figma style link is re-established 🎉
> 
> Finishing touches / todos: The test for whether the token paint equals the node's paint (comparing each RGB value) actually doesn't work at the moment 😞 It seems Figma uses a different method to convert web RGB values in the [0, 255] range to Figma's own RGB values in the [0,1] range. As far as I can see it ought to be just multiplying or dividing by 255 (which is what the [Figma Plugin Helpers](https://github.com/figma-plugin-helper-functions) is doing, but that doesn't produce the same result as Figma's 🤷🏻 I've opened an issue here regarding that: https://forum.figma.com/t/figmas-rgb-color-values-seems-to-have-a-rounding-error/23540
> 
> Until that settles, I'm doing the comparison using hex instead which produce the expected results. For simplicity I've just added that directly in the `src/plugin/setValuesOnNode.ts`file. My questions is: Do you wan't me to update the `isPaintEqual` method to use hex comparison instead or we wait for a response from Figma?
> 
> I need to finalize that for gradients as well, if we wanna go that route...
> 
> Hope we can land this in the not too distant future 😊 🤞🏻

Keep in mind the `next` branch is not the one that contains the changes required for the Themes feature. We're still working on it, it's [in this PR](https://github.com/six7/figma-tokens/pull/778) currently that once properly tested will land in the `next` branch again. 

I'll

**Comment by jakobe** [comment]
Ah ok my bad, I just assumed it was on the `next` branch.
Roger that, I'll keep an eye on that PR then 👍🏻
